DM-24375: Fix time precision issue in type decorator

andy-slac opened the pull request:
Precision of a Python float type is not enough to guarantee nanosecond
precision, need more careful handling when converting between astropy
time and nanoseconds.

andy-slac committed:
Fix time precision issue in type decorator (DM-24375)

Precision of a Python float type is not enough to guarantee nanosecond
precision, need more careful handling when converting between astropy
time and nanoseconds.

timj commented on python/lsst/daf/butler/core/ddl.py:
```diff
@@ -170,15 +170,25 @@ def process_bind_param(self, value, dialect):
+        # Special care needed to preserve nanosecond precision.
+        # Usually jd1 has no fractional part but just in case.
+        jd1, extra_jd2 = divmod(delta.jd1, 1)
+        nsec_per_day = 1_000_000_000 * 24 * 3600
```
Can we pull this out as a file constant rather than defining it in both routines?

andy-slac replied:
Moved everything to a separate module instead 🙂

timj commented on tests/test_ddl.py:
```diff
@@ -0,0 +1,99 @@
+        """Test precision of round-trip conversion.
+        """
+        # do tests at random points between epoch and max. time
+        times = [
```
Can you add some times that use nanosecond precision and also some that are very near the end of a day and not just the start?

timj reviewed (approved):
Looks great. A couple of minor comments.

andy-slac committed:
Minor refactoring for time-related methods.

Move time conversion methods from `ddl` to new module `time_utils`, add
time comparison with precision. Should make it more reusable.

timj commented on python/lsst/daf/butler/core/time_utils.py:
```diff
@@ -0,0 +1,140 @@
+        nanosecond which is larger that round-trip error for conversion
+        to/from integer nanoseconds.
+    """
+    # To compare we need them in common scale, for simlicity just
```
typo: simlicity

timj commented on tests/test_time_utils.py:
```diff
@@ -0,0 +1,125 @@
+        self.assertFalse(time_utils.times_equal(time1, time2, .25))
+        self.assertFalse(time_utils.times_equal(time2, time1, .25))
+
+        # 1/2 mksec difference
```
typo: mksec

andy-slac replied:
Do you prefer μsec? 🙂

timj replied:
I've never seen mk used for micro before so I think I'd prefer (microsec) -- but of course a μ should work fine in the source code so I'm also fine with that...
